checker-util/src/main/java/org/checkerframework/checker/regex/util/RegexUtil.java

```diff
@@ -334,4 +427,159 @@ private static String regexErrorMessage(String s, int expectedGroups, int actual
+
+        final int length = regexp.length();
+        for (int i = 0; i < length; i++) {
+            if (regexp.charAt(i) == '(') {
+                boolean isCapturingGroup = false;
+                if ((i < length - 1)
+                        && (regexp.startsWith("?<", i + 1) || regexp.charAt(i + 1) != '?')) {
+                    group += 1;
+                    unclosedCapturingGroups.push(group);
+                    isCapturingGroup = true;
+                }
+                lastUnclosedWasCapturingGroup.push(isCapturingGroup);
+                if (isCapturingGroup) {
+                    if (i > 0 && regexp.charAt(i - 1) == '|') {
+                        nonNullGroups.remove(Integer.valueOf(group));
+                    }
+                }
+            } else if (regexp.charAt(i) == ')') {
+                boolean closesCapturingGroup = lastUnclosedWasCapturingGroup.pop();
+                if (closesCapturingGroup) {
+                    Integer closedGroupIndex = unclosedCapturingGroups.pop();
+                    if ((i < length - 1 && "?*|".contains(String.valueOf(regexp.charAt(i + 1))))
+                            || (i < length - 2 && regexp.startsWith("{0", i + 1))) {
+                        nonNullGroups.remove(closedGroupIndex);
+                    }
+                }
+            } else if (regexp.charAt(i) == '[') {
+                int balance = 1, j;
+                // the loop starts from i+2 because the character class cannot be empty. "[]]" is a
+                // valid regex.
+                for (j = i + 1; j < length && balance > 0; j++) {
+                    if (regexp.charAt(j) == '[') {
+                        balance += 1;
+                    } else if (regexp.charAt(j) == ']') {
+                        // in "[]]", the first ']' is considered literally.
+                        // Similarly in "[[]]], the first ']' is considered literally.
+                        // in "[][]]]", the first and second ']' are considered literally.
+                        // Thus ']' that immediately follows a '[' does not close a character class.
+                        if (regexp.charAt(j - 1) != '[') {
+                            balance -= 1;
+                        }
+                    } else if (regexp.charAt(j) == '\\') {
+                        j = resumeFromHere(regexp, j);
+                    }
+                }
+                i = j - 1;
+            } else if (regexp.charAt(i) == '\\') {
+                i = resumeFromHere(regexp, i);
+            }
+        }
+        return nonNullGroups;
+    }
+
+    /**
+     * Returns the index till which the regex can be skipped, when '\' is encountered.
+     *
+     * @param regexp the regular expression to analyse
+     * @param st the index of the '\' which causes the skip
+     * @return the index till which the traversal can be skipped
+     */
+    private static int resumeFromHere(String regexp, int st) {
```

Comment:
This is not a very descriptive name.  (Readers don't know what is being resumed.)  Please rename the method to be more descriptive.